### PR TITLE
Add Parallel.ForEachAsync

### LINKED ISSUE #1946: Async parallel foreach
There are situations where a asynchronous parallel foreach is needed instead of the already existing ```Parallel.ForEach()``` and while it's not rocket science to implement yourself I think it would be a nice feature to be implemented in corefx. This should be seen as the asynchronous alternative to the existing ```Parallel.ForEach()```. And should take ```IEnumerable<T>```  as well as ```IAsyncEnumerable<T>```.

There are examples on this issue already here is one: https://devblogs.microsoft.com/pfxteam/implementing-a-simple-foreachasync-part-2/

```c#
        /// <summary>
        ///     Executes a foreach asynchronously.
        /// </summary>
        /// <typeparam name="T"></typeparam>
        /// <param name="source">The source.</param>
        /// <param name="dop">The degrees of parallelism.</param>
        /// <param name="body">The body.</param>
        /// <returns></returns>
        public static Task ForEachAsync<T>(this IEnumerable<T> source, int dop, Func<T, Task> body)
        {
            return Task.WhenAll(
                from partition in System.Collections.Concurrent.Partitioner.Create(source).GetPartitions(dop)
                select Task.Run(async delegate
                {
                    using (partition)
                    {
                        while (partition.MoveNext())
                            await body(partition.Current);
                    }
                }));
        }
```
### Proposed API

```c#
namespace System.Threading.Tasks
{
    public static class Parallel
    {
        public static Task ForEachAsync<T>(IAsyncEnumerable<T> source, Func<T, ValuteTask> body);
        public static Task ForEachAsync<T>(IAsyncEnumerable<T> source, ParallelOptions parallelOptions, Func<T, CancellationToken, ValueTask> body);

        public static Task ForEachAsync<T>(IEnumerable<T> source, Func<T, ValuteTask> body);
        public static Task ForEachAsync<T>(IEnumerable<T> source, ParallelOptions parallelOptions, Func<T, CancellationToken, ValuteTask> body);
    }
}
```

### Closed questions
While I consider these questions to be closed feel free to continue discussion on them. 

- Should DOP be automatic or manual at all times? 
    DOP Will have a default, just like parallel linq. However documentation should make it clear that this default can be wrong in various cases. Such as detailed by @GSPP in various comments. 
- Should it accept IEnumerable<T>?
    Yes it should 
- Should the partitioner concept be supported at all?
    No we should ensure the specified DOP is adhered to exactly. Creating batches can lead to a smaller than expected effective DOP. 
- Should we process the source items sequentially or make no attempt? 
    By default we should try to process source items sequentially to improve data locality. Reason being: HDD read performance will increase drastically. Furthermore cache usage will be improved. It should however be made very clear to people in docs, that the order cannot be guaranteed
- Should there be an alternative to Parallel.For? 
    There should be one but I'll leave that outside of this issue.
- Should the funcs return ValueTask instead of Task? 
    Yes ValueTask is more accomodating than Task (cheaper to convert Task to ValueTask than the other way arround). Furthermore it should result in a peformance increase.
- Should TaskCreationOptions.DenyChildAttach be specified? 
    Yes this should be specified see [This comment](https://github.com/dotnet/runtime/issues/1946#issuecomment-576902625) for more info
- Should there be an overload without CancellationToken in the Func? 
    No there shouldn't be. It would have little to no benefit, but would cause a lot of implications (ambiguity, analysis becomes harder)

### PR DESCRIPTION
**Pull request by stephentoub**
Fixes https://github.com/dotnet/runtime/issues/1946
cc: @kouvel, @tarekgh, @tomesendam

### CODE REVIEW

File: src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.ForEachAsync.cs [hunk 1]
```diff
@@ -0,0 +1,542 @@
...
+            {
+                var state = (AsyncForEachAsyncState<TSource>)o!;
+                bool launchedNext = false;
+
+#pragma warning disable CA2007 // Explicitly don't use ConfigureAwait, as we want to perform all work on the specified scheduler that's now current
+                try
+                {
+                    // Continue to loop while there are more elements to be processed.
+                    while (!state.Cancellation.IsCancellationRequested)
+                        {
```
Comment (gfoidl) [inline src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.ForEachAsync.cs:279]: Nit

```suggestion
                    while (!state.Cancellation.IsCancellationRequested)
                    {
```

File: src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.ForEachAsync.cs [hunk 2]
```diff
@@ -0,0 +1,542 @@
...
+namespace System.Threading.Tasks
+{
+    public static partial class Parallel
+    {
+        /// <summary>Executes a for each operation on an <see cref="System.Collections.Generic.IEnumerable{TSource}"/> in which iterations may run in parallel.</summary>
+        /// <typeparam name="TSource">The type of the data in the source.</typeparam>
+        /// <param name="source">An enumerable data source.</param>
+        /// <param name="body">An asynchronous delegate that is invoked once per element in the data source.</param>
+        /// <exception cref="System.ArgumentNullException">The exception that is thrown when the <paramref name="source"/> argument or <paramref name="body"/> argument is null.</exception>
+        /// <returns>A task that represents the entire for each operation.</returns>
```
Comment (tarekgh) [inline src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.ForEachAsync.cs:16]: >represents [](start = 33, length = 10)

could you please put a comment in all new methods specifying the DOP value will be used with every overload?

File: src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.ForEachAsync.cs [hunk 3]
```diff
@@ -0,0 +1,542 @@
...
+            Func<object?, Task> taskBody = static async o =>
+            {
+                var state = (SyncForEachAsyncState<TSource>)o!;
+                bool launchedNext = false;
+
+#pragma warning disable CA2007 // Explicitly don't use ConfigureAwait, as we want to perform all work on the specified scheduler that's now current
+                try
+                {
+                    // Continue to loop while there are more elements to be processed.
+                    while (!state.Cancellation.IsCancellationRequested)
```
Comment (tarekgh) [inline src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.ForEachAsync.cs:109]: >state [](start = 28, length = 5)

state can't be null?

**Review by tarekgh — APPROVED** [review]

File: src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.ForEachAsync.cs (hunk 2 above)
Comment (stephentoub) [reply src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.ForEachAsync.cs:16]: Sure

File: src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.ForEachAsync.cs (hunk 3 above)
Comment (stephentoub) [reply src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.ForEachAsync.cs:109]: Nope.  It's always set to 'this'.

**Comment by theodorzoulias** [comment]
Hi all. I played a bit with the `Parallel.ForEachAsync` method, and I think I may have found a bug in the current implementation. It seems that in the presence of an ambient `SynchronizationContext`, it is possible for the suppled `body` to not run always on the specified `ParallelOptions.TaskScheduler`. I observed this behavior with a `ForEachAsync` configured with a `CurrentThreadTaskScheduler` (from the [ParallelExtensionsExtras](https://devblogs.microsoft.com/pfxteam/parallelextensionsextras-tour-7-additional-taskschedulers/)), running under a [`Nito.AsyncEx.AsyncContext`](http://dotnetapis.com/pkg/Nito.AsyncEx/4.0.1/net45/doc/Nito.AsyncEx.AsyncContext). Only the first elements are invoked on the specified scheduler, and the rest are invoked on the `ThreadPoolTaskScheduler`.

Something strange also happens with this specific scheduler (`CurrentThreadTaskScheduler`), irrespective of the presence of a `SynchronizationContext`. The first few items of the enumerable are processed in reverse order (#4, #3, #2, #1, #5...). AFAICS this doesn't happen when the existing synchronous `Parallel.ForEach` method is configured with the same scheduler.

Thanks for the good work on this very much needed API!

**Comment by stephentoub** [comment]
Can you please open an issue with a repro? Thanks.

**Comment by stephentoub** [comment]
Oh, CurrentThreadTaskScheduler. If there's a sync ctx on the current thread,  the awaits in the implementation will see it.  That's expected, and there's little that can or should be done about that.  That sample CurrentThreadTaskScheduler does not make sense to use with anything on Parallel.

**Comment by theodorzoulias** [comment]
What I find compelling with the idea of combining `ForEachAsync` and `CurrentThreadTaskScheduler` is that I don't have to worry about thread synchronization inside the asynchronous delegate (the `body`). This doesn't really need to be parallelized, because it just returns a `Task`. The time-consuming work happens while the task is awaited, which is why parallelism is needed. I guess that instead of using a `CurrentThreadTaskScheduler` I could use an `ConcurrentExclusiveSchedulerPair.ExclusiveScheduler`, right? As I see now, this one seems to not be affected by the presence of the `AsyncContext`.

**Comment by theodorzoulias** [comment]
No, the `ConcurrentExclusiveSchedulerPair.ExclusiveScheduler` doesn't cut it because it doesn't preserve the `SynchronizationContext`. What I probably need is to configure the `ForEachAsync` with the `TaskScheduler.FromCurrentSynchronizationContext()`. As I see now this preserves the context, although curiously the `TaskScheduler.Current` property still shows `ThreadPoolTaskScheduler` for all but the first four elements.

About posting a repro, I am not familiar with how GitHub works, but I could post something on dotnetfiddle.net if you want.

**Comment by stephentoub** [comment]
What are you trying to achieve? I'm not understanding why you're using ForEachAsync at all.

(And fwiw, CurrentThreadTaskScheduler is definitely not what you want, regardless of sync ctx. All it does is run a queued work item synchronously instead of queueing it. It doesn't run a continuation somehow back on the original thread.)

**Comment by theodorzoulias** [comment]
This a hypothetical usage example of the `Parallel.ForEachAsync` method in a WinForms application:

    private async void Button_Click(object sender, EventArgs e)
    {
        List<string> urls = GetUrls();
        await Parallel.ForEachAsync(urls, new ParallelOptions()
        {
            TaskScheduler = TaskScheduler.FromCurrentSynchronizationContext()
        }, async (url, token) =>
        {
            Debug.WriteLine(TaskScheduler.Current);
            var document = await _httpClient.GetStringAsync(url);
            ListBox1.Items.Add($"{url}: {document.Length} chars");
        });
    }

It seems to work correctly with both a `CurrentThreadTaskScheduler` and a `TaskScheduler.FromCurrentSynchronizationContext()`. The `ListBox1` is accessed from the UI thread, as it should. It's just a bit confuzing that the `TaskScheduler.Current` property inside the asynchronous lambda returns an object of type `ThreadPoolTaskScheduler` for all but the first few urls.

**Comment by stephentoub** [comment]
Ah, you're trying to overlap the awaits but run all work on the UI thread. 

TaskScheduler and SychronizationContext are both effectively schedulers, so when there's work to be queued,  really only one can be used. If what you want is for all scheduling to be back to the original SynchronizationContext, try using TaskScheduler.FromCurrentSynchronizationContext.

**Comment by theodorzoulias** [comment]
Yes, that's the idea. From my experience in StackOverflow the last couple of years, most folks who are searching for an async-friendly version of `Parallel.ForEach` are trying to do things like that. From what I've seen most of them are not going to bother with configuring the `TaskScheduler`. They're just going to `Invoke` their way to the UI thread, using the `Dispatcher`/`Control.InvokeRequired` from inside the asynchronous lambda. 😃

Thanks for the insight about the "conflict" between the `TaskScheduler` and the `SychronizationContext`. It has not caused me any problems in the past, so I guess it's not something serious to worry about.

**Pull request merged**